### PR TITLE
Add Issue-Specific CI

### PR DESCRIPTION
**Pull request by armleo**
Partial solution for #892 . The DEF pin locations should now be applied.

 Current implementation does not properly handle multipins with the same name, like in caravel user analog project (which is my current target, to get working). Currently working on implementing this, to solve def_test issue.

@donn I don't understand on how to include the def_test to regression designs. Any suggestions?

### CODE REVIEW

File: run_designs.py
Comment (donn) [inline run_designs.py:35]: Why?

File: scripts/apply_def_template.py
Comment (donn) [inline scripts/apply_def_template.py:62]: Why not just `subprocess.check_call` if you want to spit out the output anyway?

Additionally, defutil.py is not supposed to generate much in terms of output…

File: scripts/defutil.py
Comment (donn) [inline scripts/defutil.py:44]: No commented out giant test code blocks please.

File: scripts/defutil.py
Comment (donn) [inline scripts/defutil.py:174]: `click` idiomatically uses a docstring to print help for things.

File: scripts/tcl_commands/floorplan.tcl
Comment (donn) [inline scripts/tcl_commands/floorplan.tcl:257]: Don't mix long and short flags in the same invocation.

**Review by donn — CHANGES_REQUESTED** [review]
Looks okay, just have a couple concerns:

**Comment by donn** [comment]
> @donn I don't understand on how to include the def_test to regression designs. Any suggestions?

You *can* add it to `.github/test_sets/fastest_test_set`, but it's going to compare benchmarks. Benchmarks that won't exist.

I think we need a separate CI step for things that aren't just running an entire design.

**Comment by armleo** [comment]
Will do, both suggestions and the CI.

File: run_designs.py
Comment (armleo) [reply run_designs.py:35]: If -j is not passed, then run designs errors out cryptic message stating that empty string can't be converted to integer in line 118. Maybe should just set it to one.

File: scripts/apply_def_template.py
Comment (armleo) [reply scripts/apply_def_template.py:62]: Will replace with check_call.

I think we should output the log, as this implementation, while solving def pin locations issue, will crash when strange input is supplied (like broken def). The log will help us debug this kind of issues. But I think, I will add an option to write it to some logfile instead of stdout.

File: scripts/defutil.py
Comment (armleo) [reply scripts/defutil.py:44]: The first line for ODB Reader should be removed.

Second line with openroad, I think I will leave, as it suggests a way to run this script properly, without launching the full flow.

File: scripts/tcl_commands/floorplan.tcl
Comment (armleo) [reply scripts/tcl_commands/floorplan.tcl:257]: Will do

File: scripts/defutil.py
Comment (armleo) [reply scripts/defutil.py:174]: Will fix.

File: run_designs.py
Comment (donn) [reply run_designs.py:35]: That's more reasonable, yeah.

File: scripts/apply_def_template.py
Comment (donn) [reply scripts/apply_def_template.py:62]: Makes sense.

File: scripts/defutil.py
Comment (donn) [reply scripts/defutil.py:44]: Add it to the docstring. Not in a comment.

**Comment by armleo** [comment]
Will work on this soon. Also will fix this issue: #912 . If I forget, please remind me.

**Comment by donn** [comment]
Reminder.

**Comment by armleo** [comment]
Please don't merge this yet.

I am trying to figure out a way how to: Properly both log a warning in the log file AND put a warning in the run's stdout. Whatever I do, the stdout does not appear at the output of the run.

I will also add the issue regression flow CI scripts + will add the infrastructure to do error testing. Like to test if missaligned pin will put an error I want it to put.

**Comment by armleo** [comment]
Currently:

Gives warning for missaligned pins, because of #912 .
Fixed #892 as the apply def template has been rewritten
Adds a issue_892_def_test test case
Adds issue_912_* test case. Which should (as expected) fail with warning.

Right now: Working on run_issue_regressions.py, which will run the issue_912_ test design and then run issue_912_ issue_regression_checks (not hooks). This will allow to test for the log to contain the warning and also test for return code to be not zero ( == not successful), as the error has been raised.

**Comment by armleo** [comment]
Currently, I am stuck, as I don't know how to configure GitHub CI. This is ready to be pulled, I will add the GitHub CI next Sunday.

**Comment by armleo** [comment]
Resolves following issues:
#912 warning is working
#892 def pin locations are not being applied
#881 Issue regression flow

Once CI finishes successfully, this can be merged

**Comment by armleo** [comment]
Okay. I fixed the CI issue. This pull request is ready and can be merged. Let me know if there is things you want to be changed.

**Review by donn — APPROVED** [review]
This is really, really good work. Thank you.

**Pull request merged**
